mesonbuild/linkers/linkers.py

```diff
@@ -700,14 +700,32 @@ def build_rpath_args(self, env: 'Environment', build_dir: str, from_dir: str,
         return (args, rpath_dirs_to_remove)
 
     def get_win_subsystem_args(self, value: str) -> T.List[str]:
-        if 'windows' in value:
-            args = ['--subsystem,windows']
-        elif 'console' in value:
-            args = ['--subsystem,console']
-        else:
-            raise MesonException(f'Only "windows" and "console" are supported for win_subsystem with MinGW, not "{value}".')
+        # MinGW only directly supports a couple of the possible
+        # PE application types. The raw integer works as an argument
+        # as well, and is always accepted, so we manually map the
+        # other types here. List of all types:
+        # https://docs.microsoft.com/en-us/cpp/build/reference/subsystem-specify-subsystem
+        subsystems = {
+            "native": "1",
+            "windows": "windows",
+            "console": "console",
+            "posix": "7",
+            "efi_application": "10",
+            "efi_boot_service_driver": "11",
+            "efi_runtime_driver": "12",
+            "efi_rom": "13",
+            "boot_application": "16",
+        }
+        versionsuffix = None
         if ',' in value:
-            args[-1] = args[-1] + ':' + value.split(',')[1]
+            value, versionsuffix = value.split(',', 1)
+        newvalue = subsystems.get(value)
+        if newvalue is not None:
+            if versionsuffix is not None:
+                newvalue += f',{versionsuffix}'
+            args = [f'--subsystem,{newvalue}']
+        else:
+            raise MesonException(f'Only {set(subsystems)} are supported for win_subsystem with MinGW, not "{value}".')
```

Comment:
Since we now map all types that are supported in the interpreter's validator (i.e. checked immediately when processing the `executable()` kwargs, not dependent on which linker is used), we can get rid of this MesonException as it cannot be reached.

Alternatively, we can change it to a `mesonlib.MesonBugException(f'win_subsystem: {value!r} not handled in MinGW linker. This should not be possible.')`

MesonBugException will generate a clever error message that includes the statement "this is a Meson bug and must be reported", thus encouraging people to report it as a bug because clearly the mapping dict needs to be updated to re-sync with the interpreter's validator.

Reply:
I actually ran into one of these same `MesonBugException`s while playing with the program and it helped me track down what I did wrong, so I am going to go with that option to help future developers.